akka-actor/src/main/scala/akka/io/TcpConnection.scala

```diff
@@ -257,32 +267,37 @@ private[io] abstract class TcpConnection(val tcp: TcpExt, val channel: SocketCha
     else PeerClosed
 
   def handleClose(info: ConnectionInfo, closeCommander: Option[ActorRef],
-                  closedEvent: ConnectionClosed): Unit = closedEvent match {
-    case Aborted ⇒
-      if (TraceLogging) log.debug("Got Abort command. RESETing connection.")
-      doCloseConnection(info.handler, closeCommander, closedEvent)
-    case PeerClosed if info.keepOpenOnPeerClosed ⇒
-      // report that peer closed the connection
-      info.handler ! PeerClosed
-      // used to check if peer already closed its side later
-      peerClosed = true
-      context.become(peerSentEOF(info))
-    case _ if writePending ⇒ // finish writing first
-      if (TraceLogging) log.debug("Got Close command but write is still pending.")
-      context.become(closingWithPendingWrite(info, closeCommander, closedEvent))
-    case ConfirmedClosed ⇒ // shutdown output and wait for confirmation
-      if (TraceLogging) log.debug("Got ConfirmedClose command, sending FIN.")
-
-      // If peer closed first, the socket is now fully closed.
-      // Also, if shutdownOutput threw an exception we expect this to be an indication
-      // that the peer closed first or concurrently with this code running.
-      // also see http://bugs.sun.com/view_bug.do?bug_id=4516760
-      if (peerClosed || !safeShutdownOutput())
+                  closedEvent: ConnectionClosed): Unit = {
+    // Our registered actor is now free to terminate cleanly
+    if (!(closedEvent.isConfirmed || closedEvent.isPeerClosed)) unsignDeathPact()
```

Comment:
Written negatively like this, I find it hard to understand when it applies. What happens if `handleClose` is called several times? It's a bit hard to trace.

Couldn't this more succinctly be added further down in the `case _ if writePending =>` body? That's the only case that should matter here, right? The other possibility would be just gracefully handling (= ignoring) the `Terminated` in the `closingWithPendingWrite` state.